database/database.go

```diff
@@ -1,127 +1,585 @@
+		}
+		if rowsAffected, _ := res.RowsAffected(); rowsAffected == 0 {
+			return errors.New("At StoreQuote: Could not find specified entry for overwrite")
+		}
+
+		// try to find corresponding entry in localDatabase and overwrite it
+		err = overwriteQuoteInLocalDatabase(q)
+		if err != nil {
+			return errors.New("At StoreQuote: " + err.Error())
+		}
 	}
 
 	return nil
 }
 
-// GetTeachers is a function to get teachers from the database
-func GetTeachers() ([]Teacher, error) {
+// DeleteQuote deletes the quote corresponding to the given ID from the database and the quotes slice
+// It will also modifiy the words map
+func DeleteQuote(ID int) {
+
+}
+
+/* -------------------------------------------------------------------------- */
+/*                         EXPORTED TEACHERS FUNCTIONS                        */
+/* -------------------------------------------------------------------------- */
+
+// GetTeachers returns a slice containing all teachers
+// The returned slice is not sorted
+func GetTeachers() *[]TeacherT {
+	localDatabase.mux.LockRead()
+	defer localDatabase.mux.UnlockRead()
+	teacherSlice := localDatabase.teacherSlice
+
+	return &teacherSlice
+}
 
-	rows, err := database.Query("SELECT id,name,note,title FROM teachers")
-	defer rows.Close()
+// StoreTeacher stores a new teacher
+// If the ID is not zero, StoreTeacher will try to find the corresponding teacher and overwrite it
+// If the ID is nil a new teacher will be created
+func StoreTeacher(t TeacherT) error {
+	var err error
 
+	// Verify connection to PostgreSQL database
+	err = postgresDatabase.Ping()
 	if err != nil {
-		log.Print("Cannot get teachers: ", err)
-		return nil, err
+		postgresDatabase.Close()
+		return errors.New("At StoreTeacher: " + err.Error())
 	}
 
-	var teachers []Teacher
+	if t.TeacherID == 0 {
+		// add teacher to postgresDatabase
+		err = postgresDatabase.QueryRow(
+			`INSERT INTO teachers (Name, Title, Note) VALUES ($1, $2, $3) RETURNING TeacherID`,
+			t.Name, t.Title, t.Note).Scan(&t.TeacherID)
+		if err != nil {
+			return errors.New("At StoreTeacher: " + err.Error())
+		}
 
-	for rows.Next() {
-		t := Teacher{}
-		rows.Scan(&t.ID, &t.Name, &t.Note, &t.Title)
+		// add teacher to localDatabase
+		addTeacherToLocalDatabase(t)
+	} else {
+		// try to find corresponding entry postgresDatabase and overwrite it
+		var res sql.Result
+		res, err = postgresDatabase.Exec(
+			`UPDATE teachers SET Name=$2, Title=$3, Note=$4 WHERE TeacherID=$1`,
+			t.TeacherID, t.Name, t.Title, t.Note)
+		if err != nil {
+			return errors.New("At StoreTeacher: " + err.Error())
+		}
+		if rowsAffected, _ := res.RowsAffected(); rowsAffected != 0 {
+			return errors.New("At StoreTeacher: Could not find specified entry for overwrite")
+		}
 
-		teachers = append(teachers, t)
+		// try to find corresponding entry in localDatabase and overwrite it
+		err = overwriteTeacherInLocalDatabase(t)
+		if err != nil {
+			return errors.New("At StoreTeacher: " + err.Error())
```

Comment:
is this really a fatal error? this may lead the caller to assume that the change didn't make it to the database, causing them to let the user try again, while only the local cache was the problem, though it shall not be (it's only a cache, after all, and can be reloaded quickly).